[1.20.4/5] Improved ModMismatchDisconnectedScreen and clientside handling of channel mismatches in general

RedstoneDubstep opened the pull request:
This PR aims to improve the user experience of getting disconnected from a NeoForge server due to channel mismatches, mainly by improving the usability of the ModMismatchDisconnectedScreen (the screen that shows instead of the default DisconnectedScreen if a NeoForge server disconnects the client due to channel mismatches) and also two other related changes. The ModMismatchDisconnectedScreen was previously adapted as part of the network refactor and its changes to handling channel mismatches, and this PR finalizes these adaptations to bring back most of the design elements and advantages that existed before the network refactor.

#### Changes to the ModMismatchDisconnectedScreen:
- Re-added "Connection Lost" message at the top of the screen, to clarify why the screen is shown
- Fixed the actual disconnection message possibly going outside the screen, now it will fit within the screen's dimensions
- Re-added alternate coloring for subsequent table entries, to be able to distinguish them better
- The channel mismatch reason is now automatically appended (either on the server or on the client) by the information of what mod likely owns the mismatched channel (based on the channel's namespace), so users have an easier time finding out which mod they might need to install or remove to fix their connection issues
- The channel list now merges channel mismatch entries with the same mismatch reason component into one singular entry (by default displaying one exemplary channel id and the amount of channels with an identical reason), to reduce list repetition and length; this also means that only entries of channels that are deemed to likely be owned by the same mod (based on the channel's namespaces) get merged
- Removed the assignment of the homepage URL of the mod of which the id matches a channel's namespace to that channel, mainly due to inconsistency (the URL could only be displayed when the mod was installed on the client)
- Added an "Simplified view" option button to the screen which controls if only one (default) or all channels that have the same reason component should be listed in the screen
- The channel list now displays a maximum 30 instead of 10 entries before showing the "[xyz additional, see latest.log for the full details]" message. This was done partly to bring back parity with the MMDS before the network refactor, in which the channel list had two parts (missing channels and mismatched channels) that showed 10 entries each, and partly because I felt that if we have a scroll list, we might as well include more stuff in it (since more entries don't take up more screen space).
- Fixed too early wrapping of the "Reason" component, and increased its size to half of the list (previously that table column stored a version string, which was a lot shorter than the reason components now are, which warrants the increased width of that column)
- Renamed some variables and updated some javadoc to reflect the recent changes made to this screen during the Network Refactor
#### Other changes:
- ~~Configuration and play channels now store the id of the mod that registered them (the id is retrieved through the ModLoadingContext); this information is *not* sent over the network (to prevent this PR from potentially becoming a breaking change), and so only available on the side that the channel was registered on~~ Edit: This was decided against by maintainers
- The client will now print a warning log message for all channels that failed to connect
- Fixed error message due to the packet handler trying to parse the ModdedNetworkSetupFailedPayload like a normal packet

#### Further points to discuss:
- ~~Should we rename the current "Channel name" header to "Mod name" or something similar? This is because, in my opinion, users shouldn't concern themselves about what a channel is and thus the concept of channels should be mentioned as little as possible.~~ This proposition is no longer relevant.
- Before the network refactor, the server sent all of its installed mods to the client so the client could use these mod names and particularly other metadata like version and URL for a more informative list of errors on the disconnected screen. Is that possible/feasible to add in the new protocol?
- The channel mismatch reasons sent from the server to the client are great for modders, but can be ambiguous to inexperienced mod users. Is it worth hardcode-checking the reason components (to know whether the channel was missing or mismatching) and adding a "Suggested actions: ..." text/button somewhere that shows players more user-friendly messages?
  
#### Visual comparison between the previous and updated ModMismatchDisconnectedScreen:
(Edit: the updated MMDS's design is outdated, for the current version see further below)
![image](https://github.com/neoforged/NeoForge/assets/42962686/598d3299-aa9b-41b3-9841-a8e83c18c463)
![image](https://github.com/neoforged/NeoForge/assets/42962686/82e37bd2-4f7d-424f-a689-b7b034d22c49)
(Also of note in this example: the previous MMDS had 107 entries corresponding to the 107 mismatching channels, of which only 10 were actually displayed, while the updated MMDS has 5 entries corresponding to the 5 mismatching mods.)

RedstoneDubstep committed:
Improved ModMismatchDisconnectedScreen and clientside handling of channel mismatches in general

RedstoneDubstep committed:
Ran the spotless task thingy

marchermans commented on src/main/java/net/neoforged/neoforge/client/gui/ModMismatchDisconnectedScreen.java:
```diff
@@ -114,7 +126,41 @@ record Row(MutableComponent name, MutableComponent reason) {}
+         *         and all other quasi-duplicate channels then increment the associated repetition count that is mapped to that first channel id.
+         *         Finally, the (unchanged) mismatch reason (which is the same for all quasi-duplicate entries) also gets mapped to the channel id.
+         */
+        private Map<ResourceLocation, Pair<Integer, Component>> collapseChannelData(Map<ResourceLocation, Component> mismatchedChannelData) {
```
I do not like this concept.
Modders specifically asked to be able to create channels in other namespace outside of their own. So collapsing them all and then attributing this to the mod that owns the namespace is simply not something i want to support as a maintainer.

Technici4n replied:
I agree, we need to support that. Linking a namespace to a mod is not a good idea I think. On the other hand, we don't want to spam the user with 10s of channels from the same mod. Maybe we could prioritize showing different namespaces, and if we have enough room show multiple entries from the same namespace (e.g. 100 mismatched channels with 10 different namespaces in total = show 2 channels per namespace)?

marchermans replied:
From pure testing on my 1080p monitor, there is at best room for like 5 entries in this screen. As you can see from his screenshot he gets 4 entries reasonably visible.

I would rather have them all scrollable and all properly visible, then this grouping.

RedstoneDubstep replied:
- Re automatically attributing channels to mods: I understand your point, I assumed it would be fine to automatically attribute mismatching channels to the closest matching mod because the screen had already been doing that for mod URLs before my changes, I now just also made it display the mod's names using almost the same code. (So another question would be whether the URL thing should also be scrapped now, see `MMDS#toChannelNameComponent` where the mod URL is queried and, if found, put in the hover component).
- Re collapsing channels with the same namespace: The MMDS, in my opinion, should be 1. a brief overview of the mismatch (with full mismatch details in the log) that 2. more or less tells (potentially inexperienced) users what they should do to fix the mismatch. As Technici4n said, spamming the scroll list with incredibly similar entries (same namespace + same reason) is likely disadvantageous and causes unnecessary confusion regarding that, so I would really like to have at least *some* form of summarizing/collapsing these almost duplicate entries (and not just truncating the list, like it happened previously). I understand that this is not ideal from a maintainer perspective, since grouping namespaces implies that these channel mismatches are always caused by the same mod (which they might well not be), but we need to add *some* simplification for inexperienced users, and I dare to say that in 90% of cases, if for 2 channel mismatches their namespace *and* reasons are identical, they come from one problem (= one mod).
- Re showing full channel ids/entries: In my opinion, full channel ids should ideally not be displayed in something user-facing (that raw data is what the log is for, not the screen), but I guess because the reason component is usually also long and takes up more lines, we could show some exemplary channel ids (but we need to make it clear to the inexperienced user that this is only extra info, and that the namespace of the channel gives way more info than these exemplary channel ids).

So my preferred solution now (and please tell me if and how you disagree) would be to scrap trying to query the mod from the namespace, but keep the namespace grouping, and potentially show some channel ids (in different formatting) below the (more important, so maybe highlighted) channel namespace in the list.

marchermans replied:
The problem with all your three points is that they only hold/work if and only if a given namespace can only contain channels from one mod and one mod only.
AKA this: _ if for 2 channel mismatches their namespace and reasons are identical, they come from one problem (= one mod)_ simply is not true.

This is simply not true anymore. And as such making any derivative of the namespace of a channel simply does not hold. Not to find the offending mods, and not for grouping.

Technici4n replied:
Potentially we could group channels by registrar. I.e. only merge if they were registered with the same registrar and namespace.

RedstoneDubstep replied:
This would only be a partial solution, since a) the client doesn't know which mod registered channels that are e.g. mismatching because they're required on the server and missing on the client; and b) I believe any mod can request to register packets for any registrar they want (so two mods could register their packets for the same registrar id), and we'd be where we started regarding merging entries of separate mods.
Additionally, I believe with the current code, the registrar id *has* to match the channel id when registering anyways (see `ModdedPacketRegistrar#validatePayload`, last if check), so checking for the registrar, even if possible, would not give us any extra information.

RedstoneDubstep replied:
marchermans: I am aware that this is not true anymore (I mentioned that in the message you quoted from), and in an ideal world, no merging/collapsing and assuming-that-two-channels-share-a-mod would be necessary for users to understand the MMDS and fix their issues themselves. However, it is not an ideal world, and thus, we need to help users with understanding why they can't join this specific server. This means that, alongside potentially adding a "Suggested Actions" tab to the screen (which could always be done), tradeoffs need to be made between accuracy and usability. 
A 100% accurate MMDS would not collapse anything at all (because it has no right to assume that two channel mismatches share the same cause and/or are from the same mod), but is also extremely repetitive and overwhelming for the inexperienced mod user. 
A MMDS that collapses channel namespaces is not 100% accurate anymore in that regard, which is a problem, but I believe it brings the much needed benefit of being able to tell the user "see, all of these channels share this namespace and reason, so there *could* be a causal connection leading to a single mod" (which, again, is in no way required to be true, but again, I'm going to claim that it is true in at least 85% of cases, and that glossing over the remaining 15% is a worthwile tradeoff for usability).
If you want to maintain your point that the MMDS should not collapse channels based on any criteria, I can remove the channel collapsing feature, and the MMDS would be a very accurate, but not a useful screen, barely more useful than the log (which I think would be unfortunate for a screen that originally was designed with usability for inexperienced mod users in mind). However, I would be really curious what both other NF maintainers, mod devs and mod users think of this (maybe I'll ask for opinions in the #github-discussions channel of the discord).

RedstoneDubstep replied:
I will, for now, remove that channel names automatically get enhanced by the name and URL of the mod that shares the namespace, but might not even have registered the channel itself, because in the current implementation, this gives too little of a benefit (it doesn't even work for channels required on the server and missing on the client) and instead is too misleading.
Edit: Also gonna update some javadoc

marchermans commented on src/main/java/net/neoforged/neoforge/client/gui/ModMismatchDisconnectedScreen.java:
```diff
@@ -137,25 +181,31 @@ private List<Pair<FormattedCharSequence, FormattedCharSequence>> splitLineToWidt
         /**
-         * Adds a style information to the given mod name string. The style assigned to the returned component contains the color of the mod name,
-         * a hover event containing the given id, and an optional click event, which opens the homepage of mod, if present.
+         * Uses the given channel id to return a component with the name of the mod that likely owns the channel. If no such mod is found, the namespace of the channel id is used instead.
```
See my comment above, I don't really support this "merge and blame" approach in this interface. Modders are free and explicitly requested the ability to create channels in other namespaces that are not their own.

RedstoneDubstep replied:
I believe this is resolved now, since the MMDS no longer automatically searches for mods that share the same namespace as the given channel (and might not have actually registered said channel).

RedstoneDubstep committed:
Removed code that mapped a channel to the mod with the same namespace

also updated some javadoc

RedstoneDubstep committed:
Reworked merging of MMDS to be based on the mods owning given channels

- PlayRegistrations and ConfigurationRegistrations now store the id of the mod they were created from, which is retrieved via the ModLoadingContext
- This mod id is used to enhance the channel mismatch reason component to always include the mod name (this enhancing is done on the side that the channel was registered on)
- Since some of that reason component enhancing is done in the constructor of the MMDS, the logging of channel mismatch entries has been moved to there as well
- There is now also a button that toggles listing only one (default) or all channels with the same reason component

RedstoneDubstep commented:
Alright so after these recent changes, channels are now properly grouped by their owning mod. This is achieved through the owning mod always being included (= mentioned) in the channel's mismatch reason, and because mismatched channel entries only get merged if their mismatch reasons are identical (note that in the new implementation, the channel namespace gets ignored in that merging logic, only the reason is checked), channels are strictly only merged if they were registered by the same mod.
Of note: There were some internal changes made to properly store and keep track of the mod id that owns the registered channel, however these changes have all been made in classes annotated as `@ApiStatus.Internal`, so none of the changes should be breaking ones.
Additionally, there is now a new button which controls if only one (default) or all channels that have the same reason component should be listed in the screen. Below is a visual comparison between the so-called "Simplified View" being active or not. 
A re-review based on these new changes is greatly appreciated.
![Screenshot 2024-03-06 153842](https://github.com/neoforged/NeoForge/assets/42962686/e8fe84b9-3141-4281-b262-0afec6ac4b33)
![Screenshot 2024-03-06 153934](https://github.com/neoforged/NeoForge/assets/42962686/215a7225-1540-4fc0-9f75-628a99850b14)

RedstoneDubstep committed:
Merge branch '1.20.x' into mmds-improvements

RedstoneDubstep committed:
Fixed wrong reason message being logged

RedstoneDubstep committed:
Added missing param javadoc and renamed record field

Shadows-of-Fire commented on src/main/java/net/neoforged/neoforge/network/registration/ModdedPacketRegistrar.java:
```diff
@@ -56,15 +57,15 @@ public Map<ResourceLocation, PlayRegistration<?>> getPlayRegistrations() {
         play(
                 id, new PlayRegistration<>(
-                        reader, handler, version, Optional.empty(), optional));
+                        reader, handler, ModLoadingContext.get().getActiveNamespace(), version, Optional.empty(), optional));
```
New code should not be using the static `ModLoadingContext` - see https://github.com/neoforged/FancyModLoader/issues/39

Use the Namespace of the provided id for this (and all other instances).

Technici4n replied:
See other comments as we're going back and forth. :P

RedstoneDubstep replied:
Hello, thank you for your review! I also originally considered not using the `ModLoadingContext`. Instead, I proposed changing the way the currently existing `modId` field of `ModdedPacketRegistrar` is handled, so that it's no longer required to match the namespaces of all channels registered through that registrar (since mods are allowed to register channels for any namespace they want anyways), so that I could use the `modId` field to reliaby query the id of the mod that registered the channels. However, when I suggested this, I was told to use the `ModLoadingContext` for retrieving the current mod id as it requires less changes. (For more details and also other reasons why the mentioned `modId` field should be refactored like this, see [this message in #neoforge-github](https://discord.com/channels/313125603924639766/852298000042164244/1214236463584247819) and the following related messages.)

RedstoneDubstep committed:
Merge branch '1.20.x' into mmds-improvements

RedstoneDubstep committed:
Removed usage of static ModLoadingContext, used supplied modId instead

For this solution to work, the verification that a registered payload's namespace needs to match the mod id of the registrar had to be removed, in order to allow mods passing their proper mod id to the registrar while still being able to register payloads for any namespace they desire using that registrar (as has been one of the ideas of the Network Refactor).
Suggestion: put a duplication check in NetworkRegistry#setup where the channels are built, since currently duplicate payload ids just seem to be silently overwriting each other, which doesn't feel intentional.

RedstoneDubstep commented:
Hello, I know you are currently busy with the new snapshots and (if I understood that right) an internal network api refactor by Shadows, but a continuation of the conversation above would still be greatly appreciated. With my latest commit, I've now removed all usages of the static `ModLoadingContext`, as requested by Shadows, the (in my opinion even beneficial) side effect of this change is stated in the commit's description.

Shadows-of-Fire commented:
To summarize the discussion surrounding this:

We have collectively decided the ownership-tracking approach (through the loading context) is not stable enough to warrant using, nor do we apply it for any of the other types of objects that are synchronized. We will be pursuing a new solution which better permits mods to declare their network compatibility and provide the user with actual information instead of relying solely on channel and registry compatibility.

That said, we are still interested in receiving the improvements to the Screen and the channel mismatch handling, so we can move forward here.

As far as the current change set goes, the `modid` params in `ConfigurationRegistration`, `PlayRegistration`, and `NegotiableNetworkComponent` should just be replaced with the actual registration ID (effectively making the `modid` in the `IPayloadRegistrar` defunct).  If necessary, the `ResourceLocation id` can be added as a parameter to the `*Registration` classes, but `NegotiableNetworkComponent` should need no change.

RedstoneDubstep commented:
Hello, thank you for the information! So to clarify, does that mean that no form of channel mismatch entry merging (even through tracking the owning mod of the channels) is going to get accepted (even though it has significant usability benefits compared to just listing all entries in my opinion)? If so, I can remove the relevant code from the PR to only make it consist of screen formatting fixes.

Shadows-of-Fire commented:
We can still do the merging. Realistically the unmerged version is mostly useless to an end user due to the volume of information.

RedstoneDubstep commented:
I agree, the unmerged information is just too vast and confusing for the general end user. However, I'm now confused about which criterion should be used to merge the channel mismatch entries, if both merging based on channel namespaces as well as merging based on owning mods (through mod ownership tracking) is off the table. Is there another option that allows for a usable and logical merging of channel mismatch entries that I have missed?

Shadows-of-Fire commented:
Grouping by the registered namespace should be sufficient.  This screen should basically be a last-effort fallback (same for the registry entry mismatch), as we will be expecting that mods providing networked components (registry entries, channels, data maps, etc) use the new system.

The new pipeline will be something akin to the following:
1. A declarative network version range, checked automatically as the first step of the connection.
2. Invocation of custom network compatibility checkers (for advanced handling beyond declarative version ranges).
3. The original checks, if no mods provide an explicit reason for the disconnect.

RedstoneDubstep commented:
Alright, thank you for your explanation, I don't think I completely understand it (since I haven't fully kept track of maintainer discussions regarding these changes), but I will re-add grouping of channel mismatch entries based on channel id namespaces. Should I also re-add the logic that tries to find (and display the name of) the mod of which the mod id is the same as the merged channel namespace?

Shadows-of-Fire commented:
Yeah we can retain the modid->name lookup feature.

RedstoneDubstep committed:
Re-added prior logic of grouping mismatching channels by namespace

- The name of the mod which id matches the mismatching channel namespace is displayed in the channel mismatch entry's reason component, so users have a potential mod candidate to add/update/remove
- The feature of tracking the mod that registered a channel has been removed again
- Re-added the exception that occurs when a channel gets registered to a registrar with a mod id which is different to the channel's namespace

RedstoneDubstep commented:
Alright, the prior channel-namespace-merging logic has been restored. The visual appearance of the screen doesn't differ from the previous screenshot sent into here, so I'll not post another screenshot this time. I'll resolve the open review above, please tell me what else needs to be changed.

RedstoneDubstep committed:
Merge remote-tracking branch 'origin/1.20.x' into mmds-improvements

RedstoneDubstep committed:
Fixed merge renmant and incorrect blur applied to MMDS

RedstoneDubstep commented:
Hello, I've now resolved the merge conflicts with the 1.20.5 NeoForge changes and updated some code to make the MMDS work in 1.20.5 (the screen looks really cool with the main menu panorama background). Since Shadow's Network Refactor is now merged and there are (hopefully) no more major changes to the network architecture, I'd really like to know if/how to move forward with this PR. Since this PR does not introduce a breaking change, it doesn't need to be merged during the beta window, so if you have other priorities (which you probably do) those can always go first.

RedstoneDubstep committed:
Merge branch '1.20.x' into mmds-improvements

marchermans committed:
Merge branch '1.20.x' into mmds-improvements
